caffe2/operators/batch_matmul_op.h

```diff
@@ -65,29 +74,32 @@ class BatchMatMulOp final : public Operator<Context> {
     }
 
     // Y = A * B
-    auto a_offset = A.size() / A.dim(0);
-    auto b_offset = B.size() / B.dim(0);
-    auto y_offset = a_dim0 * b_dim1;
-    for (int i = 0; i < A.dim32(0); ++i) {
-      math::Gemm<T, Context, Engine>(
-          trans_a_ ? CblasTrans : CblasNoTrans,
-          trans_b_ ? CblasTrans : CblasNoTrans,
-          a_dim0,
-          b_dim1,
-          a_dim1,
-          1,
-          A.template data<T>() + a_offset * i,
-          B.template data<T>() + b_offset * i,
-          0,
-          Y->template mutable_data<T>() + y_offset * i,
-          &context_);
-    }
+    math::GemmBatched<T, Context, Engine>(
+        trans_a_ ? CblasTrans : CblasNoTrans,
+        trans_b_ ? CblasTrans : CblasNoTrans,
+        A.size(),
+        A.dim32(0),
+        B.size(),
+        B.dim32(0),
+        a_dim0, // M
+        b_dim1, // N
+        a_dim1, // K
+        1,
+        A.template data<T>(),
+        B.template data<T>(),
+        0,
+        Y->template mutable_data<T>(),
+        &context_,
+        use_scratch_ ? scratch_ : nullptr);
     return true;
   }
 
  protected:
   bool trans_a_;
   bool trans_b_;
+
+  bool use_scratch_;
+  Tensor<Context>* scratch_;
```

Comment:
better to use a non-pointer here, or shared/unique_ptr, as this is not cleared otherwise in destructor

Reply:
Done in https://github.com/caffe2/caffe2/pull/1175